components/eam/src/chemistry/utils/tracer_data.F90

```diff
@@ -1602,6 +1637,171 @@ subroutine read_za_trc( fid, vid, loc_arr, strt, cnt, file, order )
+                !!surface padding
+                wrk2d(:,file%nlev)=wrksrf
+                deallocate(wrksrf)
+                else
+                !!padding for derivative terms
+                wrksrf=0.0_r8
+                endif
+                !!
+    endif
+    !!
+    if(associated(wrk2d_in)) then
+       wrk2d_in = reshape( wrk2d(:,:),(/file%nlat,file%nlev/), order=order )
+       deallocate(wrk2d)
+    else
+       wrk2d_in => wrk2d
+    end if
+
+    do c=begchunk,endchunk
+       ncols = get_ncols_p(c)
+       call get_rlat_all_p(c, pcols, to_lats)
+       call lininterp_init(file%lats, file%nlat, to_lats, ncols, 1, lat_wgts)
+       do k=1,file%nlev
+          call lininterp(wrk2d_in(:,k), file%nlat, wrk(1:ncols), ncols, lat_wgts)
+          loc_arr(1:ncols,k,c-begchunk+1) = wrk(1:ncols)
+       end do
+       call lininterp_finish(lat_wgts)
+    end do
+
+    if(allocated(wrk2d)) then
+       deallocate(wrk2d)
+    else
+       deallocate(wrk2d_in)
+    end if
+    !!
+    if(allocated(wrksrf)) then
+       deallocate(wrksrf)
+    end if
+!    if(dycore_is('LR')) call polar_average(loc_arr)
+  end subroutine read_za_trc_linoz
+
+!------------------------------------------------------------------------
+  subroutine read_zasrf_trc_linoz( fid, vid, loc_arr, strt, cnt, file)
+    use interpolate_data, only : lininterp_init, lininterp, interp_type, lininterp_finish
+    use phys_grid,        only : pcols, begchunk, endchunk, get_ncols_p, get_rlat_all_p, get_rlon_all_p
+    !!
+    implicit none
+    type(file_desc_t), intent(in) :: fid
+    type(var_desc_t),  intent(in) :: vid
+    integer,           intent(in) :: strt(:), cnt(:)
+    real(r8),          intent(out):: loc_arr(:,:,:)
+    type (trfile),     intent(in) :: file
+    !!
+    real(r8), allocatable, target :: wrk(:)
+    real(r8), pointer :: wrk_in(:)
+    type(interp_type) :: lat_wgts
+    real(r8) :: to_lats(pcols), to_lons(pcols)
+    integer :: c, k, ierr, ncols
+    integer :: cnt_srf(2)
+    integer :: strt_srf(2)
+    !!
+    nullify(wrk_in)
+    allocate( wrk(cnt(1)), stat=ierr )
+    if( ierr /= 0 ) then
+       write(iulog,*) 'read_zasrf_trc_linoz: wrk allocation error = ',ierr
+       call endrun
+    end if
+    !!
+    cnt_srf(1)=cnt(1)
+    cnt_srf(2)=cnt(3)
+    strt_srf(1)=strt(1)
+    strt_srf(2)=strt(3)
+    !!
+    !for surface variable with the dimension of (time,lat)
+        ierr = pio_get_var( fid, vid, strt_srf, cnt_srf, wrk )
+        !!
+        if(associated(wrk_in)) then
+                wrk_in = reshape( wrk(:),(/file%nlat/))
+                deallocate(wrk)
+        else
+                wrk_in => wrk
+        end if
+        !!
+        do c=begchunk,endchunk
+           ncols = get_ncols_p(c)
+           call get_rlat_all_p(c, pcols, to_lats)
+           call lininterp_init(file%lats, file%nlat, to_lats, ncols, 1, lat_wgts)
+           do k=1,1
+           call lininterp(wrk_in(:), file%nlat, wrk(1:ncols), ncols, lat_wgts)
```

Comment:
This is problematic, as wrk_in and wrk point to the same address. Inside lininterp (lininterp1d), within the loop computing for wrk, the original values (in wrk_in or wrk) are successively overridden. In addition to the results being unintended, because even the source data are replaced by the computed values, which are then used for computing values at remaining column, the results become threading and pe-layout sensitive.

Reply:
Hi, Wuyin. This looks like it. We will discuss a fix tomorrow.